### DISCUSSION TITLE
Local keys are now random letters/symbols

### DISCUSSION THREAD

**Discussion by uzlonewolf** [jasonacox/tinytuya#341] [General]
Just added a new bulb and it got a local key of `=fU{R=MA^tC%>>zH`.  I don't think this will affect anything as long as they don't include `"` or `\` in their character set, but it is interesting that they're moving away from hex digits.

So far I've seen the characters 
```
!#%&'()*+-./:;<=>?@[]^_`{|}~$0123456789aAbBcCdDeEfFgGhHiIjJkKlLmMnNoOpPqQrRsStTuUvVwWxXyYzZ
```

**Comment by jasonacox** [comment]
Oh wow...  No issues with tinytuya?

**Reply by uzlonewolf** [reply]
I'm not expecting (and haven't seen) any as we just pass it as a string as-is.  I was a little concerned about slashes and/or quotes in devices.json getting mangled by the JSON parser, but I'm probably over thinking it.

**Reply by jasonacox** [reply]
I can see the problem on users of the library too... the quote (`"`) may be obvious to them to escape, but if they just paste in the localkey:

```python
localkey = "=fU{R\nA^tC%>>zH"
```

becomes
```
=fU{R
A^tC%>>zH
```

I'll play around with it a bit to see what combinations I see.

**Reply by johnnykang** [reply]
i was noticing the key has changed from hex to incl special chars. and for some reason, the key (with the new format) i retrieved from the api is not longer worked with localtuya.

Could it be related ?

**Reply by jasonacox** [reply]
Hi @johnnykang  - can you post just the special characters that you are seeing (you don't have to post the whole key)?  Also, are you still able to use tinytuya (e.g. `python3 -m tinytuya scan`)?

**Reply by johnnykang** [reply]
the key format is
`G-qxg&_i!-6tj[00`

`python3 -m tinytuya scan` won't return result from my machine because all the tuya devices are on different VLan and my main PC is on different VLAN. I can try to create a LXC on the same VLAN and run that. but i am confident i will get some result.

I believe the CLI works but the new key value has issues with the other lib. I have not looked into the source code to investigate the decode logic, i am sure you have quicker and clearer idea if the new key format would work on other lib (such as localtuya).

**Reply by johnnykang** [reply]
i believe it can be easily replicated as well. 

You may try to 'reactivate' (reset the device and re-add it to the smartlife app) and you should get the new key format.

**Reply by jasonacox** [reply]
Yes, but I haven't found a case that doesn't work with tinytuya yet.

Do your VLANs connect (you can talk to the other one but UDP traffic doesn't transfer)?  If so, and you have a `devices.json` file, you can do a force scan on your IoT VLAN:  `python3 -m tinytuya -force scan`

**Reply by jasonacox** [reply]
>the key format is
G-qxg&_i!-6tj[00

I'm not sure how that would cause a problem. I don't know how localtuya does parsing but if you are using `devices.json` and python's json parser, it should be fine.

I just had a key generated that contains a single quote (`'`) which could cause an issue if it was entered in a single quoted string. The JSON payload uses double quoted (`"`) strings so it hasn't been an issue. I suspect Tuya will avoid (or escape) the double quote, but if anyone sees anything different, please raise it.

**Reply by johnnykang** [reply]
I did a force scan and indeed i can see the device that uses the new key format in the list, and also it exists in the devices.json.

i will write up some quick test python script with tinytuya and see if it can talk to the device using the new key format. If it can communicate with the device without issue, my money is on the localtuya the troublemaker.

**Reply by uzlonewolf** [reply]
If they're on a different network you can also do `python3 -m tinytuya scan -force 192.168.2.0/24 192.168.3.0/24` etc.

So far I've seen the characters 
```
'*#.={^>%()-_/`+]-~;
```

**Reply by johnnykang** [reply]
alright. 

It turned out to be the LocalTuya is the trouble. 

I've sorted out the bug on that library locally, and I am able to use the new key format without issue with HomeAssistant.

Thank you guys for helping me to identify the root cause. 

🙏

**Reply by uzlonewolf** [reply]
That's great!

What did you have to change to make it work?

**Reply by johnnykang** [reply]
with HA 2023.5, `async_get_registry` method is removed and LocalTuya has not updated the code to use `async_get` instead. 

I was getting `could not decrypt data: wrong local_key?` error. Apparently that error was not obvious in the log and it was misleading.  it was actually caused by the above exception and it has nothing to do with the actual device key.

**Reply by stu531** [reply]
Hi, I'd had a problem with this - but followed the update and got it solved. Thanks so much @johnnykang.

**Comment by szupi-ipuzs** [comment]
I've got this key:
`?:vSi=XBXH<e(o@0`

And now i have another that apart from letters (both capital and not) and numbers contains these chars: /`$:[&(
